File: week4/pa4-ner/java/FeatureFactory.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.json.JSONException;
import org.json.JSONObject;

public class FeatureFactory {

    /** Add any necessary initialization steps for your features here.
     *  Using this constructor is optional. Depending on your
     *  features, you may not need to intialize anything.
     */
    public FeatureFactory() {

    }

    /**
     * Words is a list of the words in the entire corpus, previousLabel is the label
     * for position-1 (or O if it's the start of a new sentence), and position
     * is the word you are adding features for. PreviousLabel must be the
     * only label that is visible to this method. 
     */
    private List<String> computeFeatures(List<String> words,
    		String previousLabel, int position) {

    	List<String> features = new ArrayList<String>();

    	String currentWord = words.get(position);

    	// Baseline Features  
    	features.add("word=" + currentWord);
    	features.add("prevLabel=" + previousLabel);
    	features.add("word=" + currentWord + ", prevLabel=" + previousLabel);
    	/** Warning: If you encounter "line search failure" error when
    	 *  running the program, considering putting the baseline features
    	 *  back. It occurs when the features are too sparse. Once you have
    	 *  added enough features, take out the features that you don't need. 
    	 */

    	// TODO: Add your features here
    	if ((position-2) >= 0) {
    		features.add("prevBigram=" + words.get(position-2) + ' ' + words.get(position-1));
    	}
    	if ((position-1) >= 0) {
    		features.add("bigram=" + words.get(position-1) + ' ' + currentWord);
    	}
        features.add("word=" + currentWord + ", prevLabel=" + previousLabel);
    	
    	String[] greetings = {"hi", "hello", "hallo", "hullo", "thanks", "cheers", "dear"};
    	for (String g : greetings) {
            features.add("word=" + currentWord + ",prevWord=" + g);
            features.add("word=" + currentWord + ",prevWord=" + g + ", ");
    	}
    	String[] prepositions = {"to","for","in","on","at","by","near","besides","beside","opposite","from", "of", "off", "about", "onto", "below", "under", "before", "after"}; 
    	for (String prep : prepositions) {
    		features.add("word=" + currentWord + ",prevWord=" + prep);
    	}

    	features.add(getCharsFeature(currentWord));

    	return features;
    }

	public String getCharsFeature(String currentWord) {
		String charsFeature;
		char[] chars = currentWord.toCharArray();
    	boolean firstLetterCapital = isCapital(chars[0]);
    	boolean allCapitals = true;
    	boolean mixedCapitals = false;
    	boolean hasNumeric = false;
    	boolean allNumeric = true;
    	boolean weirdStuff = false;
    	
    	for (int i=0;i<chars.length;i++) {
    		allNumeric &= isNumeric(chars[i]);
    		hasNumeric |= isNumeric(chars[i]);
    		if (!isCapital(chars[i])) {
    			allCapitals = false;
    		}
    		if (i>0) {
    			mixedCapitals |= isCapital(chars[i]);
    		}
    		weirdStuff |= !isAlphaNumeric(chars[i]);
    	}
    	if (weirdStuff) {
    		charsFeature = "case=Weird";
    	}
    	else if (allNumeric) {
    		charsFeature = "case=Numeric";
    	}
    	else if (hasNumeric) {
    		charsFeature = "case=MixedNumeric";
    	}
    	else if (allCapitals) {
    		charsFeature = "case=Capitals";
    	}
    	else if (mixedCapitals) {
    		charsFeature = "case=Mixed";
    	}
    	else if (firstLetterCapital) {
    		charsFeature = "case=Title";
    	}
    	else {
    		charsFeature = "case=Lower";
    	}
		return charsFeature;
	}
	
	public boolean isAlphaNumeric(char c) {
		return isNumeric(c) || isCapital(c) || isLower(c);
	}

	public boolean isNumeric(char c) {
		return c >= '0' && c <= '9';
	}
    
    public boolean isCapital(char c) {
    	return c >= 'A' && c <= 'Z';
    }
    
    public boolean isLower(char c) {
    	return c >= 'a' && c <= 'z';
    }

    /** Do not modify this method **/
    public List<Datum> readData(String filename) throws IOException {

	List<Datum> data = new ArrayList<Datum>();
	BufferedReader in = new BufferedReader(new FileReader(filename));

	for (String line = in.readLine(); line != null; line = in.readLine()) {
	    if (line.trim().length() == 0) {
		continue;
	    }
	    String[] bits = line.split("\\s+");
	    String word = bits[0];
	    String label = bits[1];

	    Datum datum = new Datum(word, label);
	    data.add(datum);
	}

	return data;
    }

    /** Do not modify this method **/
    public List<Datum> readTestData(String ch_aux) throws IOException {

	List<Datum> data = new ArrayList<Datum>();

	for (String line : ch_aux.split("\n")) {
	    if (line.trim().length() == 0) {
		continue;
	    }
	    String[] bits = line.split("\\s+");
	    String word = bits[0];
	    String label = bits[1];

	    Datum datum = new Datum(word, label);
	    data.add(datum);
	}

	return data;
    }

    /** Do not modify this method **/
    public List<Datum> setFeaturesTrain(List<Datum> data) {
	// this is so that the feature factory code doesn't accidentally use the
	// true label info
	List<Datum> newData = new ArrayList<Datum>();
	List<String> words = new ArrayList<String>();

	for (Datum datum : data) {
	    words.add(datum.word);
	}

	String previousLabel = "O";
	for (int i = 0; i < data.size(); i++) {
	    Datum datum = data.get(i);

	    Datum newDatum = new Datum(datum.word, datum.label);
	    newDatum.features = computeFeatures(words, previousLabel, i);
	    newDatum.previousLabel = previousLabel;
	    newData.add(newDatum);

	    previousLabel = datum.label;
	}

	return newData;
    }

    /** Do not modify this method **/
    public List<Datum> setFeaturesTest(List<Datum> data) {
	// this is so that the feature factory code doesn't accidentally use the
	// true label info
	List<Datum> newData = new ArrayList<Datum>();
	List<String> words = new ArrayList<String>();
	List<String> labels = new ArrayList<String>();
	Map<String, Integer> labelIndex = new HashMap<String, Integer>();

	for (Datum datum : data) {
	    words.add(datum.word);
	    if (labelIndex.containsKey(datum.label) == false) {
		labelIndex.put(datum.label, labels.size());
		labels.add(datum.label);
	    }
	}

	// compute features for all possible previous labels in advance for
	// Viterbi algorithm
	for (int i = 0; i < data.size(); i++) {
	    Datum datum = data.get(i);

	    if (i == 0) {
		String previousLabel = "O";
		datum.features = computeFeatures(words, previousLabel, i);

		Datum newDatum = new Datum(datum.word, datum.label);
		newDatum.features = computeFeatures(words, previousLabel, i);
		newDatum.previousLabel = previousLabel;
		newData.add(newDatum);

	    } else {
		for (String previousLabel : labels) {
		    datum.features = computeFeatures(words, previousLabel, i);

		    Datum newDatum = new Datum(datum.word, datum.label);
		    newDatum.features = computeFeatures(words, previousLabel, i);
		    newDatum.previousLabel = previousLabel;
		    newData.add(newDatum);
		}
	    }

	}

	return newData;
    }

    /** Do not modify this method **/
    public void writeData(List<Datum> data, String filename)
	throws IOException {


	FileWriter file = new FileWriter(filename + ".json", false);

	       
	for (int i = 0; i < data.size(); i++) {
	    try {
		JSONObject obj = new JSONObject();
		Datum datum = data.get(i);
		obj.put("_label", datum.label);
		obj.put("_word", base64encode(datum.word));
		obj.put("_prevLabel", datum.previousLabel);

		JSONObject featureObj = new JSONObject();

		List<String> features = datum.features;
		for (int j = 0; j < features.size(); j++) {
		    String feature = features.get(j).toString();
		    featureObj.put("_" + feature, feature);
		}
		obj.put("_features", featureObj);
		obj.write(file);
		file.append("\n");
	    } catch (JSONException e) {
		// TODO Auto-generated catch block
		e.printStackTrace();
	    }
	}
	file.close();
    }

    /** Do not modify this method **/
    private String base64encode(String str) {
	Base64 base = new Base64();
	byte[] strBytes = str.getBytes();
	byte[] encBytes = base.encode(strBytes);
	String encoded = new String(encBytes);
	return encoded;
    }

}

File: week4/pa4-ner/python/FeatureFactory.py
import json, sys
import base64
import re
from Datum import Datum
import nltk
from nltk.corpus import names

class FeatureFactory:
    re_is_lower_char = re.compile('/[a-z]/')
    re_is_upper_char = re.compile('/[A-Z]/')
    re_is_numeric = re.compile('/[0-9]/')


    """
    Add any necessary initialization steps for your features here
    Using this constructor is optional. Depending on your
    features, you may not need to intialize anything.
    """
    def __init__(self):
        pass


    """
    Words is a list of the words in the entire corpus, previousLabel is the label
    for position-1 (or O if it's the start of a new sentence), and position
    is the word you are adding features for. PreviousLabel must be the
    only label that is visible to this method.
    """

    def computeFeatures(self, words, previousLabel, position):
        features = []
        currentWord = words[position]

        """ Baseline Features """
        features.append("word=" + currentWord)
        features.append("prevLabel=" + previousLabel)
        features.append("word=" + currentWord + ", prevLabel=" + previousLabel)
        """
            Warning: If you encounter "line search failure" error when
            running the program, considering putting the baseline features
        back. It occurs when the features are too sparse. Once you have
            added enough features, take out the features that you don't need.
        """


        """ TODO: Add your features here
            Baseline score:
                precision = 0.801673640167364 #number selected items that are correct
                recall = 0.5229257641921398 #number correct items that are selected
                F1 = 0.6329699372315825
        """
        # precision = 0.8603736479842674
        # recall = 0.47762008733624456
        # F1 = 0.6142506142506142


        if (position-1) > 0:
            features.append("word=" + currentWord + ",prevWord=" + words[position-1])
            features.append("prevWord=" + words[position-1])
        #     features.append("bigram=" + words[position-1] + ' ' + currentWord)
        #     prepositions = ['to','for','by','besides','from','about']
        #     if words[position-1] in prepositions:
        #         features.append("word=" + currentWord + ",preposition=" + words[position-1])
        #     pospositions = ['of']
        #     if currentWord in pospositions:
        #         features.append("posposition=" + currentWord + ",prevWord=" + words[position-1])
        #     if words[position-1] in names.words():
        #         features.append("fullname=" + words[position-1] + ' ' + currentWord)
        #         features.append("surname_1=" + currentWord)
            if re.search('^[A-Za-z]\.$', words[position-1]):
                features.append("initial=" + words[position-1])
                features.append("surname_2=" + currentWord)
                features.append("fullname_2=" + words[position-1] + ' ' + currentWord)
        #     speaking = ['said', 'told', 'says', 'say', 'tells', 'tell']
        #     ignored = ['I', 'You', 'He', 'She', 'It', 'We', 'They']
        #     if words[position-1][0].isupper() and currentWord in speaking and words[position-1] not in ignored and not words[position-1].isupper():
        #         features.append("said=" + words[position-1])

        #     if words[position-1][0].isupper() and currentWord == ',':
        #         features.append("comma=" + words[position-1])
        #     if words[position-1][0].isupper() and currentWord == "'s":
        #         features.append("genitive_s=" + words[position-1])

        # if (position-2) > 0:
        #     speaking = ['said', 'told', 'says', 'say', 'tells', 'tell']
        #     ignored = ['I', 'You', 'He', 'She', 'It', 'We', 'They']
        #     punctuation = ['.', ',', ';', ':']

        #     if currentWord in punctuation:
        #         if words[position-2][0].isupper() and words[position-1] in speaking and words[position-2] not in ignored and not words[position-2].isupper():
        #             features.append("said_2=" + words[position-2])
        #         if words[position-2] in punctuation:
        #             features.append("leading_ending=" + words[position-2] + words[position-1] + currentWord)
        #     if words[position-2] == '\"' and words[position-1][0].isupper() and currentWord in speaking:
        #         features.append("said_3=" + words[position-1])
        #     if currentWord == '\"' and words[position-1][0].isupper() and words[position-2] in speaking:
        #         features.append("said_4=" + words[position-1])




        # features.append("word=" + currentWord + ", prevLabel=" + previousLabel)


        if currentWord in names.words():
            features.append("name=" + currentWord)


        #features.append(self.getCharsFeature(currentWord))
        case = ''
        titleCase = False
        if self.re_is_numeric.search(currentWord):
            case = "numeric=True"
        else:
            if currentWord.isupper():
                case = "case=Upper"
            elif currentWord[0].isupper():
                cnt = 0
                titleCase = True
                for c in currentWord:
                    if cnt > 0:
                        if c.isupper():
                            titleCase = False
                    cnt += 1
                if currentWord.startswith('O\'') or (currentWord.startswith('Mc') and currentWord[2].isupper()) or (currentWord.startswith('Mac') and currentWord[3].isupper()):
                    titleCase = True
                if titleCase:
                    case = "case=Title"
                    if previousLabel == 'PERSON':
                        features.append("surname="+currentWord)
        features.append(case)
        if (position-1) > 0:
            features.append("prevWord=" + words[position-1] + ',' + case)

        return features



    """
    getCharsFeature
    """
    def getCharsFeature(self, currentWord):
        firstLetterCapital = False
        allCapitals = True
        mixedCapitals = False
        hasNumeric = False
        allNumeric = True
        weirdStuff = False

        cnt = 0
        for c in currentWord:
            allNumeric &= self.isNumeric(c)
            hasNumeric |= self.isNumeric(c)
            if not self.isCapital(c):
                allCapitals = False
            if cnt==0:
                firstLetterCapital = self.isCapital(c)
            if (cnt>0):
                mixedCapitals |= self.isCapital(c)

            weirdStuff |= not self.isAlphaNumeric(c)
            cnt += 1

        if (weirdStuff):
            charsFeature = "case=Weird"
        elif (allNumeric):
            charsFeature = "case=Numeric"
        elif (hasNumeric):
            charsFeature = "case=MixedNumeric"
        elif (allCapitals):
            charsFeature = "case=Capitals"
        elif (mixedCapitals):
            charsFeature = "case=Mixed"
        elif (firstLetterCapital):
            charsFeature = "case=Title"
        else:
            charsFeature = "case=Lower"
        return charsFeature;

    def isAlphaNumeric(self, c):
        return self.isNumeric(c) or self.isCapital(c) or self.isLower(c)

    def isNumeric(self, c):
        return self.re_is_numeric.search(c) is not None


    def isCapital(self, c):
        return self.re_is_upper_char.search(c) is not None

    def isLower(self, c):
        return self.re_is_lower_char.search(c) is not None


    """ Do not modify this method """
    def readData(self, filename):
        data = []

        for line in open(filename, 'r'):
            line_split = line.split()
            # remove emtpy lines
            if len(line_split) < 2:
                continue
            word = line_split[0]
            label = line_split[1]

            datum = Datum(word, label)
            data.append(datum)

        return data

    """ Do not modify this method """
    def readTestData(self, ch_aux):
        data = []

        for line in ch_aux.splitlines():
            line_split = line.split()
            # remove emtpy lines
            if len(line_split) < 2:
                continue
            word = line_split[0]
            label = line_split[1]

            datum = Datum(word, label)
            data.append(datum)

        return data


    """ Do not modify this method """
    def setFeaturesTrain(self, data):
        newData = []
        words = []

        for datum in data:
            words.append(datum.word)

        ## This is so that the feature factory code doesn't
        ## accidentally use the true label info
        previousLabel = "O"
        for i in range(0, len(data)):
            datum = data[i]

            newDatum = Datum(datum.word, datum.label)
            newDatum.features = self.computeFeatures(words, previousLabel, i)
            newDatum.previousLabel = previousLabel
            newData.append(newDatum)

            previousLabel = datum.label

        return newData

    """
    Compute the features for all possible previous labels
    for Viterbi algorithm. Do not modify this method
    """
    def setFeaturesTest(self, data):
        newData = []
        words = []
        labels = []
        labelIndex = {}

        for datum in data:
            words.append(datum.word)
            if not labelIndex.has_key(datum.label):
                labelIndex[datum.label] = len(labels)
                labels.append(datum.label)

        ## This is so that the feature factory code doesn't
        ## accidentally use the true label info
        for i in range(0, len(data)):
            datum = data[i]

            if i == 0:
                previousLabel = "O"
                datum.features = self.computeFeatures(words, previousLabel, i)

                newDatum = Datum(datum.word, datum.label)
                newDatum.features = self.computeFeatures(words, previousLabel, i)
                newDatum.previousLabel = previousLabel
                newData.append(newDatum)
            else:
                for previousLabel in labels:
                    datum.features = self.computeFeatures(words, previousLabel, i)

                    newDatum = Datum(datum.word, datum.label)
                    newDatum.features = self.computeFeatures(words, previousLabel, i)
                    newDatum.previousLabel = previousLabel
                    newData.append(newDatum)

        return newData

    """
    write words, labels, and features into a json file
    Do not modify this method
    """
    def writeData(self, data, filename):
        outFile = open(filename + '.json', 'w')
        for i in range(0, len(data)):
            datum = data[i]
            jsonObj = {}
            jsonObj['_label'] = datum.label
            jsonObj['_word']= base64.b64encode(datum.word)
            jsonObj['_prevLabel'] = datum.previousLabel

            featureObj = {}
            features = datum.features
            for j in range(0, len(features)):
                feature = features[j]
                featureObj['_'+feature] = feature
            jsonObj['_features'] = featureObj

            outFile.write(json.dumps(jsonObj) + '\n')

        outFile.close()

